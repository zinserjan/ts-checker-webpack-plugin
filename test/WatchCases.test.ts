import * as path from "path";
import * as fs from "fs-extra";
import webpack = require("webpack");
import MemoryFs = require("memory-fs");
import pDefer = require("p-defer");
import { createAssertExpectation, satisfiesVersionRequirements } from "./_util/testHelper";
import * as processMock from "./_util/processMock";

const testCasesPath = path.join(__dirname, "watchCases");
const tests = fs.readdirSync(testCasesPath).filter(dir => fs.statSync(path.join(testCasesPath, dir)).isDirectory());

const tmpPath = path.resolve(__dirname, "../.tmp/watchCases");

let time = Date.now() / 1000;
const copySmart = (srcPath: string, targetPath: string) => {
  time += 10;
  const files = fs.readdirSync(srcPath);

  for (const file of files) {
    const srcFilePath = path.join(srcPath, file);
    const targetFilePath = path.join(targetPath, file);
    const isDirectory = fs.statSync(srcFilePath).isDirectory();
    if (isDirectory) {
      copySmart(srcFilePath, targetFilePath);
    } else {
      const content = fs.readFileSync(srcFilePath).toString("utf-8");
      if (/DELETE/.test(content)) {
        fs.removeSync(targetFilePath);
      } else {
        const exists = fs.existsSync(targetFilePath);
        if (exists) {
          fs.writeFileSync(targetFilePath, content);
        } else {
          fs.copySync(srcFilePath, targetFilePath);
        }
        // make sure that the last modified time was touched
        fs.utimesSync(targetFilePath, time, time);
      }
    }
  }
};

jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;

describe("WatchCases", () => {
  beforeAll(() => {
    return fs.emptyDir(tmpPath);
  });

  beforeEach(() => processMock.register());
  afterEach(() => processMock.unregister());

  tests.forEach(testName => {
    const testPath = path.join(testCasesPath, testName);
    const tmpTestPath = path.join(tmpPath, testName);
    const tmpStepsPath = path.join(tmpTestPath, "steps");
    const srcTarget = path.join(tmpTestPath, "src");
    const webpackConfigPath = path.join(tmpTestPath, "webpack.config.ts");
    const skipTest = !satisfiesVersionRequirements(path.join(testCasesPath, "versions.json"));
    const assertExpectation = createAssertExpectation(path.join(testPath, "expectation.ts"));

    (skipTest ? it.skip : it)(testName, async () => {
      await fs.copy(testPath, tmpTestPath);

      const copyDiff = async (stepIndex: string) => {
        const srcStep = path.join(tmpStepsPath, stepIndex);
        await copySmart(srcStep, srcTarget);
      };

      const steps = (await fs.readdir(tmpStepsPath)).sort();
      const stepCount = steps.length;
      const firstStep = steps.shift() as string;
      await copyDiff(firstStep);

      const webpackConfig = require(webpackConfigPath);
      const c = webpack(webpackConfig);

      const compilers = (c as any).compilers ? (c as any).compilers : [c];
      compilers.forEach((compiler: any) => {
        const ifs = compiler.inputFileSystem;
        compiler.inputFileSystem = Object.create(ifs);
        compiler.outputFileSystem = new MemoryFs();
        compiler.inputFileSystem.readFile = function() {
          const args = Array.prototype.slice.call(arguments);
          const callback = args.pop();
          ifs.readFile.apply(
            ifs,
            args.concat([
              (err: any, result: any) => {
                if (err) return callback(err);
                callback(null, result.toString("utf-8").replace(/\r/g, ""));
              },
            ])
          );
        };
        compiler.apply(new webpack.optimize.OccurrenceOrderPlugin(false));
      });

      const deferred = pDefer();
      const watching = c.watch(
        {
          aggregateTimeout: 1000,
          poll: 400,
        },
        (err, stats) => {
          if (err) {
            return deferred.reject(err);
          }

          assertExpectation(stats, stepCount - (steps.length + 1));

          if (steps.length > 0) {
            const nextStep = steps.shift() as string;
            copyDiff(nextStep);
          } else {
            watching.close(() => null);
            process.nextTick(deferred.resolve);
          }
        }
      );
      return deferred.promise;
    });
  });
});
