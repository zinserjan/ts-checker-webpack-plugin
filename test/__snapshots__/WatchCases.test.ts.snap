// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchCases basic-type-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,7): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases basic-type-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases basic-type-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases basic-type-no-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases check-file-until-error-fixed 1`] = `
Object {
  "errors": Array [
    "xfile
(1,7): error TS2322",
    "xfile
(1,7): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases check-file-until-error-fixed 2`] = `
Object {
  "errors": Array [
    "xfile
(1,7): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases check-file-until-error-fixed 3`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases check-interfaces-until-fixed 1`] = `
Object {
  "errors": Array [
    "xfile
(4,14): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases check-interfaces-until-fixed 2`] = `
Object {
  "errors": Array [
    "xfile
(3,14): error TS2420",
    "xfile
(4,14): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases check-interfaces-until-fixed 3`] = `
Object {
  "errors": Array [
    "xfile
(3,14): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases check-interfaces-until-fixed 4`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases check-interfaces-until-fixed 5`] = `
Object {
  "errors": Array [
    "xfile
(3,14): error TS2420",
    "xfile
(4,14): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases class-hierarchy-types 1`] = `
Object {
  "errors": Array [
    "xfile
(3,14): error TS2420",
    "xfile
(7,18): error TS2345",
    "xfile
(4,11): error TS2345",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases class-hierarchy-types 2`] = `
Object {
  "errors": Array [
    "xfile
(7,18): error TS2345",
    "xfile
(4,11): error TS2345",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases class-hierarchy-types 3`] = `
Object {
  "errors": Array [
    "xfile
(7,18): error TS2345",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases class-hierarchy-types 4`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-error 1`] = `
Object {
  "errors": Array [
    "xfile
(3,20): error TS2339",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-no-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-removal-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases css-modules-removal-no-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases file-removal-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases file-removal-no-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ignore-child-compiler-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ignore-child-compiler-no-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 1`] = `
Object {
  "errors": Array [
    "xfile
(4,20): error TS2307",
    "xfile
(1,20): error TS2307",
    "xfile
(1,20): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 2`] = `
Object {
  "errors": Array [
    "xfile
(4,20): error TS2307",
    "xfile
(1,20): error TS2307",
    "xfile
(1,20): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 3`] = `
Object {
  "errors": Array [
    "xfile
(4,20): error TS2307",
    "xfile
(1,20): error TS2307",
    "xfile
(1,20): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 4`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 5`] = `
Object {
  "errors": Array [
    "xfile
(2,20): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 6`] = `
Object {
  "errors": Array [
    "xfile
(4,20): error TS2307",
    "xfile
(1,20): error TS2307",
    "xfile
(1,20): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases module-extension-type-definition 7`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases node-module-with-type-definition 1`] = `
Object {
  "errors": Array [
    "xfile
(2,17): error TS2345",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases node-module-with-type-definition 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases react-components 1`] = `
Object {
  "errors": Array [
    "xfile
(11,17): error TS2339",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases react-components 2`] = `
Object {
  "errors": Array [
    "xfile
(11,17): error TS2339",
    "xfile
(9,9): error TS2322",
    "xfile
(9,9): error TS2322",
    "xfile
(9,9): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases react-components 3`] = `
Object {
  "errors": Array [
    "xfile
(11,17): error TS2339",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases skipLibCheck-false 1`] = `
Object {
  "errors": Array [
    "xfile
(1,1): error TS1046",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases skipLibCheck-false 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases skipLibCheck-false 3`] = `
Object {
  "errors": Array [
    "xfile
(1,17): error TS2345",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases skipLibCheck-false 4`] = `
Object {
  "errors": Array [
    "xfile
(1,17): error TS2345",
    "xfile
(1,1): error TS1046",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-error 2`] = `
Object {
  "errors": Array [
    "xfile
(1,22): error TS2306",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-error 3`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-error 4`] = `
Object {
  "errors": Array [
    "xfile
(3,22): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-2-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-2-error 2`] = `
Object {
  "errors": Array [
    "xfile
(1,22): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-2-error 3`] = `
Object {
  "errors": Array [
    "xfile
(3,22): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-error 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases ts-dependency-only-removal-error 3`] = `
Object {
  "errors": Array [
    "xfile
(1,22): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases tslint-error 1`] = `
Object {
  "errors": Array [
    "xfile
  2:11  error  ' should be \\"  quotemark
",
    "xfile
  4:11  error  ' should be \\"  quotemark
",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases tslint-error 2`] = `
Object {
  "errors": Array [
    "xfile
  4:11  error  ' should be \\"  quotemark
",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases tslint-error 3`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases tslint-warning 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [
    "xfile
  2:11  warning  ' should be \\"  quotemark
",
    "xfile
  4:11  warning  ' should be \\"  quotemark
",
  ],
}
`;

exports[`WatchCases tslint-warning 2`] = `
Object {
  "errors": Array [],
  "warnings": Array [
    "xfile
  4:11  warning  ' should be \\"  quotemark
",
  ],
}
`;

exports[`WatchCases tslint-warning 3`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases type-patches-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`WatchCases type-patches-error 2`] = `
Object {
  "errors": Array [
    "xfile
(1,22): error TS2307",
  ],
  "warnings": Array [],
}
`;

exports[`WatchCases type-patches-error 3`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;
